tree: correctly format the "unknown" oid for pg compatibility

DrewKimball opened the pull request:
The `REGTYPE`, `REGPROC`, `REGCLASS`, and `REGNAMESPACE` types are all
in the oid type family, but unlike `OID`, display `-` instead of `0` when
the value is zero. Previously, this was handled by setting the `name` field
of `DOid` to `-`, but the handling was incomplete. This commit removes the
old special cases, and instead adds special handling to the functions that
parse an oid from a string, as well as the `DOid.Format()` method. This way,
correctly formatting the "unknown" oid doesn't rely on remembering to set
the `name` field.

In addition, this commit modifies some of the random datum generation
logic to create types other than the canonical `T_oid`, like `T_regrole`.

Fixes #118273

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/118587)
<!-- Reviewable:end -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice find!!

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-:-Np_rSk96AZcIDu8c2Ba:bsf1ej1)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-:-NpfDCZN2dX_5idtpfIX:bnfp4nl)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

I ended up having to make some changes to the places where we generate datums for tests, as well.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-:-Nq3Y7xqCYDP1fn-i-Am:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
tree: correctly format the "unknown" oid for pg compatibility

The `REGTYPE`, `REGPROC`, `REGCLASS`, and `REGNAMESPACE` types are all
in the oid type family, but unlike `OID`, display `-` instead of `0` when
the value is zero. Previously, this was handled by setting the `name` field
of `DOid` to `-`, but the handling was incomplete. This commit removes the
old special cases, and instead adds special handling to the functions that
parse an oid from a string, as well as the `DOid.Format()` method. This way,
correctly formatting the "unknown" oid doesn't rely on remembering to set
the `name` field.

In addition, this commit modifies some of the random datum generation
logic to create types other than the canonical `T_oid`, like `T_regrole`.

Fixes #118273

Release note: None

DrewKimball commented:
After wrestling with tests for a bit, I've changed the approach. I removed the old (incomplete) handling of "unknown" oids. We now only output the `-` when either printing text results to the client, or when casting to a string. There are also two locations where we parse an oid from a string; those correctly handle `-` as well.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 5 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-:-Nrq71nY0bS-3Kchul6o:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @rytaft)

___
*[`pkg/sql/logictest/testdata/logic_test/pgoidtype` line 702 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-Nrq6wifCK3Q1XRy1A0S:-Nrq6wifCK3Q1XRy1A0T:bs6oyul) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10edcbf82277c65c27697e6da794dedd3c1bdb6/pkg/sql/logictest/testdata/logic_test/pgoidtype#L702)):*
> ```
> INSERT INTO t118273 VALUES (0, 0);
> 
> skipif config local-mixed-23.1
> ```

nit: I think you can mention multiple configs to skip on a single line.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
nice!

> Reviewed 1 of 5 files at r2, 12 of 12 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-:-NrqAdhY0QtZd_1GirJo:b-7m74ox)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 2 stale) (waiting on @DrewKimball, @mgartner, @michae2, and @rytaft)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/118587)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @mgartner, @michae2, @rafiss, @rytaft, and @yuzefovich)

___
*[`pkg/sql/logictest/testdata/logic_test/pgoidtype` line 702 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/118587#-Nrq6wifCK3Q1XRy1A0S:-Nrwt7bXCQqUUuaSqpaD:b4m8pkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b10edcbf82277c65c27697e6da794dedd3c1bdb6/pkg/sql/logictest/testdata/logic_test/pgoidtype#L702)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I think you can mention multiple configs to skip on a single line.
</blockquote></details>

TIL, thanks.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@DrewKimball should we push it over the finish line? We just got another failure like this over in #122817

DrewKimball commented:
Sorry, I lost track of this one. I'll rebase.

DrewKimball commented:
TFTRs!

bors r+

yuzefovich commented:
blathers backport release-24.1

yuzefovich commented:
blathers backport release-23.2

mgartner commented:
blathers backport release-23.1
